// express-simple-form/classes/InputSelect.js
'use strict';

let escape = require('escape-html');

let Input = require('./Input');

class InputSelect extends Input {

	options(options) {
		this._options = options;
		return this;
	}

	selected(selected) {
		this._selected = selected;
		return this;
	}

	isMultiple() {
		return 'multiple' in this.opts;
	}

	isSelected (key) {
		var opts = this.opts;
		if (this.isMultiple())
			return (this._selected || []).indexOf(key) >= 0;
		return this._selected === key;
	}

	_restore () {
		let name;
		if (!(name = this.opts.name)) return;
		if (!(name in this.data)) return;
		this.selected(this.data[name]);
	}

	renderOptions () {
		var options = this._options || {};
		return Object.keys(options).reduce((cur, key)=>{
			 return cur += [
										'<option',
											'value="'+escape(key)+'"',
											(this.isSelected(key) ? 'selected' : ''), 
										'>',
											escape(options[key]),
										'</option>'
									].join('\r\n');
		},'');
	}

	render() {

		this._restore();
		
		return [
			'<select',
				this.renderAttrs(),
			'>',
			this.renderOptions(),
			'</select>'
		].join(' ')
	}

}

module.exports = InputSelect;
